MOxUnit/assertExceptionThrown.m

```diff
@@ -1,63 +1,120 @@
+% SYNTAX:
+%   assertExceptionThrown(func, <expectedID>, <message>)
 %
-% assertExceptionThrown(func,expected_id[,msg])
+% PURPOSE:
+%   This function allows one to test for exceptions being thrown, and
+%   optionally, pass a custome message in response to a failure.
 %
-% Inputs:
-%   func            function handle that should take no inputs and should
-%                   raise an exception
-%   expected_id     the idenitifier of the exception that func should
-%                   raise
-%   msg             optional custom message
+%   e.g.
+%   % Assert that sin will throw when its argument is a struct
+%   >> assertExceptionThrown( @()sin( struct([]) ) )
 %
-% Throws:
-%   'moxunit:exceptionNotRaised'    func() does not raise an exception
-%   'moxunit:wrongExceptionRaised'  func() does raise an exception but with
-%                                       an identifier different from
-%                                       expected_id
+%   % Assert that sin throws AND that the ID is 'MATLAB:UndefinedFunction'
+%   >> assertExceptionThrown( @()sin( struct([]) ), 'MATLAB:UndefinedFunction')
+%  
+% INPUT:
+%   func        - Function handle that is expected to throw, with the prototype
+%                   [varargout{:}] = func()
+%   expectedID  - Idenitifier of the expected exception.  If expectedID is
+%                 empty, then any exception (identifier) is allowed.  When 
+%                 exactly two inputs are passed an ambiguity arises between
+%                 expectedID and message because these inputs share a common
+%                 type.  By default, we resolve this ambiguity using
+%                 moxunit_util_is_message_identifier().  If any exception is
+%                 permitted, we recommend explicitly defining expectedID to be
+%                 empty.
+%                   Default: ''
+%   message     - Custom message to be included when func fails to throw
+%   
+% THROWS:
+%   'moxunit:exceptionNotRaised'    - func() does not raise an exception but was
+%                                     expected to do so.
+%   'moxunit:wrongExceptionRaised'  - func() does raise an exception but with
+%                                     an identifier different from expected_id
 %
+% ASSUMPTIONS: 
+%   All input variables are of the correct type, valid (if applicable), and 
+%   given in the correct order. 
 %
-% Raises:
-%
-    exception_was_raised=false;
-    if moxunit_util_platform_is_octave()
-        try
-            func();
-        catch
-            exception_was_raised=true;
-            [unused,found_id]=lasterr();
-        end
-    else
-        try
-            func();
-        catch mexception;
-            exception_was_raised=true;
-            found_id=mexception.identifier;
+% See also moxunit_util_is_message_identifier
+%-------------------------------------------------------------------------------
+%}
+function assertExceptionThrown(func, expectedID, message)
+
+% Parse Inputs
+switch nargin
+    case 1
```

Comment:
See me comment below.  This would be easy to support, but something needs to be done about "empty" vs implicitly defined then.   I'm thinking about it.  Good point.